src/Middleware/AjaxMiddleware.php

```diff
@@ -0,0 +1,169 @@
+<?php
+
+namespace Ajax\Middleware;
+
+use Cake\Core\Configure;
+use Cake\Core\InstanceConfigTrait;
+use Cake\Event\Event;
+use Cake\Event\EventManager;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+
+/**
+ * Ajax Middleware to respond to AJAX requests.
+ *
+ * Works together with the AjaxView to easily switch
+ * output type from HTML to JSON format. Replaces some
+ * of the component functionality when the Authorization
+ * middleware is in use.
+ *
+ * It will also avoid redirects and pass those down as content
+ * of the JSON response object.
+ *
+ * @author Greg Schmidt, with much code copied from the AjaxComponent class by Mark Scherer
+ * @license http://opensource.org/licenses/mit-license.php MIT
+ */
+class AjaxMiddleware {
+
+	use InstanceConfigTrait;
+
+	/**
+	 * @var \Cake\Controller\Controller
+	 */
+	protected $Controller;
+
+	/**
+	 * @var array
+	 */
+	protected $_defaultConfig = [
+		'viewClass' => 'Ajax.Ajax',
+		'autoDetect' => true,
+		'resolveRedirect' => true,
+		'flashKey' => 'Flash.flash',
+		'actions' => [],
+		'jsonOptions' => JSON_HEX_TAG | JSON_HEX_APOS | JSON_HEX_AMP | JSON_HEX_QUOT | JSON_PARTIAL_OUTPUT_ON_ERROR,
+	];
+
+	/**
+	 * Constructor
+	 *
+	 * @param array|string $config Array of configuration settings or string with authentication service provider name.
+	 */
+	public function __construct($config = []) {
+		$defaults = (array)Configure::read('Ajax') + $this->_defaultConfig;
+		$config += $defaults;
+		$this->setConfig($config);
+	}
+
+	/**
+	 * Callable implementation for the middleware stack.
+	 *
+	 * @param \Psr\Http\Message\ServerRequestInterface $request The request.
+	 * @param \Psr\Http\Message\ResponseInterface $response The response.
+	 * @param callable $next The next middleware to call.
+	 * @return \Psr\Http\Message\ResponseInterface A response.
+	 */
+	public function __invoke(ServerRequestInterface $request, ResponseInterface $response, $next) {
+		$respondAsAjax = $this->_config['autoDetect'] && $this->_isActionEnabled($request) && $request->is('ajax');
+		if ($respondAsAjax) {
+			EventManager::instance()->on('Controller.beforeRender', [$this, 'beforeRender']);
+		}
+
+		$response = $next($request, $response);
+
+		if ($respondAsAjax) {
+			EventManager::instance()->off('Controller.beforeRender', [$this, 'beforeRender']);
+			if ($this->_config['resolveRedirect'] && $response->hasHeader('Location')) {
```

Comment:
Are you sure `$response->hasHeader()` method exists?
From the looks into it, only request has this method?